### PR TITLE
Refactor unnecessary try/catch

### LINKED ISSUE #3006: ConsiderEmptyDataSourceAsInconclusive throws to catch the exception on the next line
When 0 items are produced and the code does not throw, we throw just to catch the exception.  This could be made more efficient by extracting the catch and using it in `if` and `catch` paths.

https://grep.app/search?q=%26%26%20.%2AConsiderEmptyDataSourceAsInconclusive&regexp=true

Code is introduced here: https://github.com/microsoft/testfx/pull/2771/files

PR has no comments that would mention this.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #3006

### CODE REVIEW

File: src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs [hunk 1]
```diff
@@ -409,24 +409,22 @@ private static bool ProcessTestDataSourceTests(UnitTestElement test, ReflectionT
 
             // This code is to discover tests. To run the tests code is in TestMethodRunner.ExecuteDataSourceBasedTests.
             // Any change made here should be reflected in TestMethodRunner.ExecuteDataSourceBasedTests as well.
-            try
-            {
-                data = dataSource.GetData(methodInfo);
+            data = dataSource.GetData(methodInfo);
 
-                if (!data.Any())
+            if (!data.Any())
```
Comment (Youssef1313) [inline src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs:419]: This actually was NEVER true before.

GetData used to throw the exception when it's about to return empty.

**Review by Evangelink — DISMISSED** [review]

**Comment by Youssef1313** [comment]
@Evangelink I'm disabling auto-merge as I'm making changes since your last review. Once I'm able to get it green, I'll let you know to re-review.

File: src/TestFramework/TestFramework/Interfaces/EmptyDataSourceExceptionInfoExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+internal static class EmptyDataSourceExceptionInfoExtensions
+{
+    internal static ArgumentException GetExceptionForEmptyDataSource(this ITestDataSource dataSource, MethodInfo testMethodInfo)
+        => dataSource is ITestDataSourceEmptyDataSourceExceptionInfo info ? info.GetExceptionForEmptyDataSource(testMethodInfo) : new ArgumentException(
+            string.Format(
+                CultureInfo.InvariantCulture,
+                FrameworkMessages.DynamicDataIEnumerableEmpty,
+                "GetData",
+                dataSource.GetType().Name));
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Interfaces/EmptyDataSourceExceptionInfoExtensions.cs]: Not mandatory but I tend to find it easier when scanning code to have the ternary split like

```suggestion
        => dataSource is ITestDataSourceEmptyDataSourceExceptionInfo info
            ? info.GetExceptionForEmptyDataSource(testMethodInfo)
            : new ArgumentException(
                string.Format(
                    CultureInfo.InvariantCulture,
                    FrameworkMessages.DynamicDataIEnumerableEmpty,
                    "GetData",
                    dataSource.GetType().Name));
```

File: src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs [hunk 3]
```diff
@@ -409,25 +409,26 @@ private static bool ProcessTestDataSourceTests(UnitTestElement test, ReflectionT
 
             // This code is to discover tests. To run the tests code is in TestMethodRunner.ExecuteDataSourceBasedTests.
             // Any change made here should be reflected in TestMethodRunner.ExecuteDataSourceBasedTests as well.
-            try
+            data = dataSource.GetData(methodInfo);
+
+            if (!data.Any())
             {
-                data = dataSource.GetData(methodInfo);
+                if (MSTestSettings.CurrentSettings.ConsiderEmptyDataSourceAsInconclusive)
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs]: Optional nit: I usually to prefer exit conditions first to reduce mental overhead so I would suggest to revert the if to throw and not to have some else.

File: src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs [hunk 4]
```diff
@@ -311,25 +311,26 @@ private bool ExecuteDataSourceBasedTests(List<TestResult> results)
...
-                        dataSource = testDataSource.GetData(_testMethodInfo.MethodInfo);
 
-                        if (!dataSource.Any())
+                    // This code is to execute tests. To discover the tests code is in AssemblyEnumerator.ProcessTestDataSourceTests.
+                    // Any change made here should be reflected in AssemblyEnumerator.ProcessTestDataSourceTests as well.
+                    dataSource = testDataSource.GetData(_testMethodInfo.MethodInfo);
+
+                    if (!dataSource.Any())
+                    {
+                        if (MSTestSettings.CurrentSettings.ConsiderEmptyDataSourceAsInconclusive)
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs]: Optional nit: I usually to prefer exit conditions first to reduce mental overhead so I would suggest to revert the if to throw and not to have some else.

**Review by Evangelink — APPROVED** [review]
Small nits but LGTM, thanks @Youssef1313

**Pull request merged**